File: nids_blackhole.cpp
// nids_blackhole.cpp
#include <pcap.h>
#include <map>
#include <csignal>
#include <cstdlib>

std::map<std::string,int> counts;
const int THRESH = 100;

void block_ip(const std::string &ip) {
    std::string cmd = "iptables -A INPUT -s " + ip + " -j DROP";
    system(cmd.c_str());
}

void packet_handler(u_char*, const struct pcap_pkthdr* h, const u_char* bytes) {
    if (h->caplen < 34) return;
    const u_char *ip = bytes + 26;
    char src[16];
    snprintf(src, sizeof(src), "%u.%u.%u.%u", ip[12], ip[13], ip[14], ip[15]);
    auto &cnt = counts[src];
    if (++cnt == THRESH) block_ip(src);
}

int main(){
    char errbuf[PCAP_ERRBUF_SIZE];
    pcap_t *handle = pcap_open_live("eth0", BUFSIZ, 1, 1000, errbuf);
    if (!handle) { perror(errbuf); return 1; }
    pcap_loop(handle, 0, packet_handler, nullptr);
    return 0;
}

File: nids_blackhole.py
# nids_blackhole.py
import subprocess
from scapy.all import sniff, IP

THRESHOLD = 10
counts = {}

def block_ip(ip):
    subprocess.run(['sudo', 'iptables', '-A', 'INPUT', '-s', ip, '-j DROP'])
    print(f"👎 Blackholed {ip}")

def monitor(pkt):
    if IP in pkt:
        src = pkt[IP].src
        counts[src] = counts.get(src, 0) + 1
        if counts[src] == THRESHOLD:
            block_ip(src)

if __name__ == "__main__":
    print("Starting NIDS...")
    sniff(prn=monitor, store=False)
